File: growth/oj/leet_code/algorithms/338-counting-bits.py
# -*- coding: utf-8 -*-

class Solution(object):
    def countBits(self, num):
        """
        :type num: int
        :rtype: List[int]
        """

        res = [0, 1, 1]  # [0, 1, 2]
        if num <= 2:
            return res[:num+1]
        last_power_2 = 2
        i = 3
        while i <= num:
            if i != 0 and i & (i - 1) == 0:
                last_power_2 = i
                res.append(1)
            else:
                idx = i % last_power_2
                res.append(res[idx] + res[last_power_2])
            i += 1
        return res
    def normalCountBits(self, num):
        res = [0, 1]
        i = 2
        while i <= num:
            cnt = 0
            handle_i = i
            while handle_i != 0:
                handle_i = handle_i & (handle_i-1)
                cnt += 1
            res.append(cnt)
            i += 1
        return res






if __name__ == '__main__':
    obj = Solution()
    for i in xrange(10):
        print obj.countBits(i)
        print obj.normalCountBits(i)
        print "-----------\n"
    '''
    print obj.countBits(2)
    print obj.countBits(5)
    print obj.countBits(10)
    r1 = obj.countBits(30)
    r2 = obj.normalCountBits(30)
    print r1
    print r2
    '''
    #num = 0
    #for i in obj.countBits(20):
    #    print "num: %d bits: %d\n" % (num, i)
    #    num += 1
    pass

File: growth/oj/leet_code/algorithms/338-counting-bits.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    vector<int> countBits(int num) {
        if (num == 0)
            return vector<int>{0};
        if (num == 1)
            return vector<int>{0, 1};
        if (num <= 2)
            return vector<int>{0,1,1};
        vector<int> res(num+1);
        res[0] = 0;
        res[1] = 1;
        res[2] = 1;
        int cur_2_power = 2;
        for (int i = 3; i <= num; i++) {
            if (i == cur_2_power * 2) {
                res[i] = 1;
                cur_2_power *= 2;
            } else {
                res[i] = 1 + res[i-cur_2_power];
            }
        }
        return res;
    }
};


int main() {
    Solution s;
    auto res = s.countBits(3);
    cout << res.size() << endl;
    return 0;
}
